code/modules/xenoarchaeology/traits/xenoartifact_majors.dm

```diff
@@ -0,0 +1,589 @@
+	log_game("[X] swapped the identities of [key_name_admin(victims[1])] & [key_name_admin(victims[2])] at [world.time]. [X] located at [X.x] [X.y] [X.z]")
+	addtimer(CALLBACK(src, .proc/undo_swap), ((X.charge*0.20) SECONDS)+ 6 SECONDS) //6 extra seconds while targets are asleep
+	X.cooldownmod = ((X.charge*0.3)SECONDS)+ 6 SECONDS
+	victims = list()
+
+/datum/xenoartifact_trait/major/mirrored/proc/swap(var/atom/victim_a, var/atom/victim_b)
+	if(QDELETED(victim_a) || QDELETED(victim_b))
+		victims = list()
+		return
+	var/mob/living/caster = victim_a
+	var/mob/living/victim = victim_b
+	if(HAS_TRAIT(caster, TRAIT_MINDSWAPPED)) //doesn't really matter which we check
+		REMOVE_TRAIT(caster, TRAIT_MINDSWAPPED, type)
+		REMOVE_TRAIT(victim, TRAIT_MINDSWAPPED, type)
+	else
+		ADD_TRAIT(caster, TRAIT_MINDSWAPPED, type)
+		ADD_TRAIT(victim, TRAIT_MINDSWAPPED, type)
+
+	var/mob/dead/observer/ghost_v = victim.ghostize(0)
+	var/mob/dead/observer/ghost_c = caster.ghostize(0)
+	ghost_v?.mind.transfer_to(caster)
+	ghost_c?.mind.transfer_to(victim)
+	if(ghost_v?.key)
+		caster?.key = ghost_v?.key
+	if(ghost_c?.key)
+		victim?.key = ghost_c?.key
+	qdel(ghost_v)
+	qdel(ghost_c)
+
+	caster.Unconscious(6 SECONDS)
+	victim.Unconscious(6 SECONDS)
+	reverse_victims += list(list(caster, victim))
+
+/datum/xenoartifact_trait/major/mirrored/proc/undo_swap()
+	for(var/list/L as() in reverse_victims)
+		if(L.len > 1)
+			//convert to atoms to check qdel
+			var/mob/living/a = L[1]
+			var/mob/living/b = L[2]
+			if(!QDELETED(a) && !QDELETED(b))
+				swap(L[1], L[2])
+	reverse_victims = list()
+
+
+///============
+/// EMP, produces an empulse
+///============
+/datum/xenoartifact_trait/major/emp
+	label_name = "EMP"
+	label_desc = "EMP: The shape of the Artifact doesn't resemble anything particularly interesting. Technology around the Artifact seems to malfunction."
+	flags = URANIUM_TRAIT
+	weight = 25 //annoying trait
+
+/datum/xenoartifact_trait/major/emp/activate(obj/item/xenoartifact/X)
+	playsound(get_turf(X), 'sound/magic/disable_tech.ogg', 50, TRUE)
+	empulse(get_turf(X.loc), max(1, X.charge*0.03), max(1, X.charge*0.07, 1)) //This might be too big
+
+///============
+/// Invisible, makes the target invisible for a short time
+///============
+/datum/xenoartifact_trait/major/invisible //One step closer to the one ring
+	label_name = "Transparent"
+	label_desc = "Transparent: The shape of the Artifact is difficult to percieve. You feel the need to call it, precious..."
+	weight = 25
+	var/list/victims = list()
+
+/datum/xenoartifact_trait/major/invisible/on_item(obj/item/xenoartifact/X, atom/user, atom/item)
+	if(istype(item, /obj/item/laser_pointer))
+		var/obj/item/laser_pointer/L = item
+		if(L.energy)
+			to_chat(user, "<span class='info'>The [item.name]'s light passes through the structure.</span>")
+			return TRUE
+	..()
+
+/datum/xenoartifact_trait/major/invisible/activate(obj/item/xenoartifact/X, mob/living/target)
+	if(isliving(target))
+		victims += target
```

Comment:
This would probably be better to use weak references here, since you only ever rarely use the references to victims